overlord/devicestate/devicestate.go

```diff
@@ -341,91 +344,255 @@ func extractDownloadInstallEdgesFromTs(ts *state.TaskSet) (firstDl, lastDl, firs
-		} else {
-			ts, err = snapstate.LinkNewBaseOrKernel(st, new.Base())
+// pass both the snap name and the model snap, as it is possible that
+// the model snap is nil for UC16 models
+type modelSnapsForRemodel struct {
+	currentSnap      string
+	currentModelSnap *asserts.ModelSnap
+	new              *asserts.Model
+	newSnap          string
+	newModelSnap     *asserts.ModelSnap
+}
+
+func remodelKernelOrBaseTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+	userID := 0
+	newModelSnapChannel, err := modelSnapChannelFromDefaultOrPinnedTrack(ms.new, ms.newModelSnap)
+	if err != nil {
+		return nil, err
+	}
+	if ms.currentSnap == ms.newSnap {
+		changed := false
+		if ms.new.Grade() != asserts.ModelGradeUnset {
+			// UC20 models can specify default channel for all snaps
+			// including base and kernel
+			// new model uses the same base or kernel
+			changed, err = installedSnapChange(st, ms.newSnap, newModelSnapChannel)
+			if err != nil {
+				return nil, err
+			}
+		} else if ms.newModelSnap != nil && ms.newModelSnap.SnapType == "kernel" {
+			// UC18 models could only specify track for the kernel
+			// and gadget snaps
+			changed = ms.currentModelSnap.PinnedTrack != ms.newModelSnap.PinnedTrack
 		}
-		if err != nil {
-			return nil, err
+		if changed {
+			// new modes specifies the same snap, but with a new channel
+			return snapstateUpdateWithDeviceContext(st, ms.newSnap,
+				&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
 		}
-		tss = append(tss, ts)
+		return nil, nil
+	}
+
+	// new model specifies a different snap
+	needsInstall, err := notInstalled(st, ms.newModelSnap.SnapName())
+	if err != nil {
+		return nil, err
+	}
+	if needsInstall {
+		// which needs to be installed
+		return snapstateInstallWithDeviceContext(ctx, st, ms.newSnap,
+			&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+			userID, snapstate.Flags{}, deviceCtx, fromChange)
 	}
-	if current.Base() != new.Base() {
-		needsInstall, err := notInstalled(st, new.Base())
+
+	if ms.new.Grade() != asserts.ModelGradeUnset {
+		// pre UC20 remodels would not update an already installed snap
+
+		// which is already installed, but may be using a different channel
+		changed, err := installedSnapChange(st, ms.newModelSnap.SnapName(), newModelSnapChannel)
 		if err != nil {
 			return nil, err
 		}
-		if needsInstall {
-			ts, err = snapstateInstallWithDeviceContext(ctx, st, new.Base(), nil, userID, snapstate.Flags{}, deviceCtx, fromChange)
+		if changed {
+			return snapstateUpdateWithDeviceContext(st, ms.newSnap,
+				&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
+		}
+	}
```

Comment:
there's still a wrinkle here, if the update procudes only a switch-snap I think we also need the link-snap as produced by LinkNewBaseOrKernel, because we still need to turn the base or kernel into the boot ones